include/tunable_block_reduction.h

```diff
@@ -22,115 +22,127 @@ namespace quda
-    {
-      const_cast<Arg &>(arg).grid_dim = tp.grid;
-      const_cast<Arg &>(arg).block_dim = tp.block;
-      launch_device<Block::block.size() - 1, Block, Functor>(arg, tp, stream);
-    }
-
-    template <int idx, typename Block, template <typename> class Functor, typename Arg>
-    std::enable_if_t<idx != 0, void> launch_host(const TuneParam &tp, const Arg &arg)
-    {
-      if (tp.block.x == Block::block[idx])
-        BlockKernel2D_host<Functor>(BlockKernelArg<Block::block[idx], Arg>(arg));
-      else
-        launch_host<idx - 1, Block, Functor>(tp, arg);
-    }
-
-    template <int idx, typename Block, template <typename> class Functor, typename Arg>
-    std::enable_if_t<idx == 0, void> launch_host(const TuneParam &tp, const Arg &arg)
-    {
-      if (tp.block.x == Block::block[idx])
-        BlockKernel2D_host<Functor>(BlockKernelArg<Block::block[idx], Arg>(arg));
-      else
-        errorQuda("Unexpected block size %d", tp.block.x);
-    }
-
-    template <template <typename> class Functor, typename Block, typename Arg>
-    void launch_host(const TuneParam &tp, const qudaStream_t &, const Arg &arg)
-    {
-      const_cast<Arg &>(arg).grid_dim = tp.grid;
-      const_cast<Arg &>(arg).block_dim = tp.block;
-      launch_host<Block::block.size() - 1, Block, Functor>(tp, arg);
-    }
-
     /**
-       @brief Launch function for BlockReductionKernel2D.
+       @brief Launch the block reduction kernel on the set loation
+       with a given block size on the device performing the block
+       reduction defined in the functor.  We recursively iterate over
+       the length of instantiated block sizes until we succeed, or
+       error out.
        @tparam Functor Class which performs any pre-reduction
        transformation (defined as ternary operator) as well as a store
        method for writing out the result.
-       @tparam Block Class that must contain a static std::array of
-       block sizes "block" we wish to instantiate
-       @param[in] tp Kernel launch parameters
-       @param[in] stream Stream in which to execute
-       @param[in,out] arg Algorithm meta data
+       @tparam Block Class that must contain a static array of block
+       sizes "block" we wish to instantiate
+       @tparam enable_host Whether to enable host compilation (default is not to)
+       @param[in] arg Kernel argument struct
+       @param[in] tp The launch parameters
+       @param[in] stream The stream on which the execution is done
      */
     template <template <typename> class Functor, typename Block, bool enable_host = false, typename Arg>
-    std::enable_if_t<!enable_host, void> launch(const TuneParam &tp, const qudaStream_t &stream, const Arg &arg)
+    void launch(const TuneParam &tp, const qudaStream_t &stream, const Arg &arg)
     {
       if (location == QUDA_CUDA_FIELD_LOCATION) {
         launch_device<Functor, Block>(tp, stream, arg);
+      } else if constexpr (enable_host) {
```

Comment:
I don't know if I like this being popped into this PR --- is there a workflow that tests this?

Reply:
not sure I follow.  This isn't a few feature.  It's just a cleanup of the prior code to use `if constexpr` instead of SFINAE.  And yes, this is used all over the place, e.g., the copy kernels.